c8d/list: Add test and combine size

vvoland opened the pull request:
**- What I did**
- Add unit test for `Images` (image list)
- Skip images which don't match platform matcher
- Extract and simplify some code
- Count containers using `ImageManifest`

**- How I did it**
See commits 🙈 

**- How to verify it**
CI

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
c8d/list: Combine size

Multi-platform images are coalesced into one entry now.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
daemon/c8d: Use i.images and i.content

Use `image.Store` and `content.Store` stored in the ImageService struct
instead of fetching it every time from containerd client.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
daemon/c8d: Cache SnapshotService

Avoid fetching `SnapshotService` from client every time. Fetch it once
and then store when creating the image service.

This also allows to pass custom snapshotter implementation for unit
testing.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
specialimage: Return optional ocispec.Index

To ease accessing image descriptors in tests.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/list: Add TestImageList

Add unit test for `Images` implementation.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/list: Count containers by their manifest

Move containers counting out of `singlePlatformImage` and count them
based on the `ImageManifest` property.

(also remove ChainIDs calculation as they're no longer used)

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/list: Simplify "best" image selection

Don't save all present images,  inline the sorting into the loop
instead.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/list: Extract imageSummary function

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/list: Remove outdated TODO

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
```bash
$ docker pull busybox
Using default tag: latest
latest: Pulling from library/busybox
c2bf9493c1bf: Download complete 
Digest: sha256:6d9ac9237a84afe1516540f40a0fafdc86859b2141954b4d643af7066d598b74
Status: Downloaded newer image for busybox:latest
docker.io/library/busybox:latest

$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED       SIZE
busybox      latest    6d9ac9237a84   5 weeks ago   6.09MB

$ docker pull --platform linux/amd64 busybox
Using default tag: latest
latest: Pulling from library/busybox
9ad63333ebc9: Download complete 
Digest: sha256:6d9ac9237a84afe1516540f40a0fafdc86859b2141954b4d643af7066d598b74
Status: Image is up to date for busybox:latest
docker.io/library/busybox:latest

$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED       SIZE
busybox      latest    6d9ac9237a84   5 weeks ago   12.7MB
```

thaJeztah commented on daemon/containerd/image_list.go:
```diff
@@ -78,7 +78,7 @@ func (i *ImageService) Images(ctx context.Context, opts imagetypes.ListOptions)
 	}
 
 	// TODO(thaJeztah): do we need to take multiple snapshotters into account? See https://github.com/moby/moby/issues/45273
-	snapshotter := i.client.SnapshotService(i.snapshotter)
+	snapshotter := i.snapshotterService
```
Hmm.. this means we won't be able to use multiple snapshotters (see the TODO comment above). If we do this, we should store it indexed by snapshotter.

thaJeztah commented on daemon/containerd/image_list.go:
```diff
@@ -257,36 +275,39 @@ func (i *ImageService) Images(ctx context.Context, opts imagetypes.ListOptions)
-		return nil, nil, errors.Wrapf(err, "failed to get rootfs of image %s", imageManifest.Name())
-	}
-
+func (i *ImageService) singlePlatformSize(ctx context.Context, imgMfst *ImageManifest) (totalSize int64, contentSize int64, _ error) {
```
This name is a bit confusing. I need to look into what exactly it returns;  `totalSize` is the size for all platforms? what is `contentSize` in this context?

thaJeztah replied:
Oh! This is for a single-arch image (so "total" for the image, in this case being the total for a single arch, correct?)

vvoland replied:
Yes, total here means both content and "unpacked" size.

thaJeztah commented on daemon/containerd/service.go:
```diff
@@ -74,6 +78,14 @@ func NewService(config ImageServiceConfig) *ImageService {
+		return s
+	}
+
+	return i.client.SnapshotService(snapshotter)
```
Should this be stored in the map, or was it intentional to only keep a reference to the default one?

thaJeztah commented on internal/testutils/specialimage/multilayer.go:
```diff
@@ -53,7 +53,7 @@ func MultiLayer(dir string) error {
 
 	legacyManifests := []manifestItem{
-		manifestItem{
+		{
```
Oh! Good that I see this; I just noticed earlier that `gofmt` gave me a diff in this file, and "go report" showed it as well; https://goreportcard.com/report/github.com/docker/docker

thaJeztah commented on daemon/containerd/image_test.go:
```diff
@@ -296,3 +298,45 @@ func newTestDB(ctx context.Context, t *testing.T) *metadata.DB {
+
+func (s *testSnapshotterService) Stat(ctx context.Context, key string) (snapshots.Info, error) {
+	return snapshots.Info{}, nil
+}
+
+func (s *testSnapshotterService) Update(ctx context.Context, info snapshots.Info, fieldpaths ...string) (snapshots.Info, error) {
+	return snapshots.Info{}, nil
+}
+
+func (s *testSnapshotterService) Usage(ctx context.Context, key string) (snapshots.Usage, error) {
+	return snapshots.Usage{}, nil
+}
+
+func (s *testSnapshotterService) Prepare(ctx context.Context, key, parent string, opt ...snapshots.Opt) ([]mount.Mount, error) {
+	return nil, nil
+}
+
+func (s *testSnapshotterService) View(ctx context.Context, key, parent string, opt ...snapshots.Opt) ([]mount.Mount, error) {
+	return nil, nil
+}
+
+func (s *testSnapshotterService) Commit(ctx context.Context, name, key string, opts ...snapshots.Opt) error {
+	return nil
+}
+
+func (s *testSnapshotterService) Remove(ctx context.Context, key string) error {
+	return nil
+}
+
+func (s *testSnapshotterService) Walk(ctx context.Context, fn snapshots.WalkFunc, filters ...string) error {
+	return nil
+}
+
+func (s *testSnapshotterService) Close() error {
+	return nil
+}
+
+func (s *testSnapshotterService) Mounts(ctx context.Context, key string) ([]mount.Mount, error) {
+	return nil, nil
+}
```
It looks like the test only calls 2 methods (`Stat` and `Usage`), so instead of implementing stubs for all methods, you could consider embedding the interface (calling any of the non-implemented ones will panic, so should surface);

```suggestion
type testSnapshotterService struct {
	snapshots.Snapshotter
}

func (s *testSnapshotterService) Stat(ctx context.Context, key string) (snapshots.Info, error) {
	return snapshots.Info{}, nil
}

func (s *testSnapshotterService) Usage(ctx context.Context, key string) (snapshots.Usage, error) {
	return snapshots.Usage{}, nil
}
```

laurazard replied:
I keep forgetting about this "trick", it's super useful though.

vvoland replied:
Yes, thanks, I always forget about it too!

thaJeztah commented on daemon/containerd/image_list.go:
```diff
@@ -218,6 +213,14 @@ func (i *ImageService) Images(ctx context.Context, opts imagetypes.ListOptions)
 
+			if opts.ContainerCount {
+				i.containers.ApplyAll(func(c *container.Container) {
+					if c.ImageManifest != nil && c.ImageManifest.Digest == img.Target().Digest {
```
For a follow-up; we should really document the `ImageManifest` field in container; I went looking to see if this was the "platform-specific" manifest, or if this was the manifest-index ("root" of the image), but the field is currently not documented.

thaJeztah commented on daemon/containerd/image_list.go:
```diff
@@ -199,6 +199,18 @@ func (i *ImageService) Images(ctx context.Context, opts imagetypes.ListOptions)
+				platform = dockerImage.Platform
+			}
+
+			if !platformMatcher.Match(platform) {
```
Perhaps would be worth adding your commit-message description as a comment here;  this would not be obvious immediately when looking at the code.

> Currently this won't have any real effect because the platform matcher matches all platform and is only used for sorting.


Curious if this _would_ filter, wouldn't that mean that `docker image ls` would not include sizes for all images?

laurazard replied:
I'd assume it would only include the size for matching-platform images. However, I think that makes sense: if the passes in a specific platform matcher, it's because they only want to know about those particular images.

thaJeztah replied:
Yeah my train of thought here was that;


```
abcd ubuntu:latest 200MB
```


Is the compact view of;

```
abcd ubuntu:latest 200MB
    |- ubuntu:latest X64   100MB
    |- ubuntu:latest ARM64 100MB
```

So when "collapsed", it would show the summary of it all.


For sure, this is a bit of a limitation of the current means we have, and to some extent there's also something to be said for that to be showing "default" platform. But "hiding" the 100MB for the non-matching platform is also surprising (It says I have 200MB used for images, but I only see 100??)

vvoland replied:
Yes, I think it makes more sense to show the total size. Fixed.

laurazard reviewed (approved):
LGTM!
Just noticed a little typo/duplicated word in `c8d/list: Count containers by their manifest`'s text:
```
Move containers counting out out `singlePlatformImage` and count them
based on the `ImageManifest` property.
```

vvoland committed:
c8d/list: Skip images with non matching platform

Currently this won't have any real effect because the platform matcher
matches all platform and is only used for sorting.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
Thanks @laurazard! This one was meant to be `out of`, not `out out` 😅

thaJeztah commented on daemon/containerd/service.go:
```diff
@@ -74,6 +78,16 @@ func NewService(config ImageServiceConfig) *ImageService {
 }
 
+func (i *ImageService) snapshotterService(snapshotter string) snapshots.Snapshotter {
+	s, ok := i.snapshotterServices[snapshotter]
+	if !ok {
+		s = i.client.SnapshotService(snapshotter)
+		i.snapshotterServices[snapshotter] = s
+	}
+
+	return s
```
For posterity / future self; I _think_ this is fine to do, but technically this map would be "unbounded", so would allow an "unlimited" number of snapshotters to be referenced here.

I don't think there's any realistic security here (unless someone would want to DOS themselves by having a gazillion snapshotters).

thaJeztah reviewed (approved):
LGTM


I wonder if (with more complexity, and more iterating needed to collect all the data) if we should benchmark some of this. Do we need to consider caching at some point? Caching is hard though (and we should not optimise if not needed); more so because the list of present images / platforms will change at any moment, so cache invalidation would now be an issue.

thaJeztah commented:
Looks like this one is flaky (I think I mentioned the same one on another PR);

```
=== FAIL: github.com/docker/docker/integration/networking TestBridgeICCWindows/User_defined_nat_network (9.20s)
    bridge_test.go:242: assertion failed: error is not nil: Post "http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.45/containers/e6bc72d3a70cd87e40bdc547f722749fb7038a26c421697cd350ebcced73db0b/start": context deadline exceeded
    panic.go:523: assertion failed: error is not nil: Error response from daemon: error while removing network: network mynat id 01daafb478bec6f00a16d87bd570a8328c8135c212575c12e29cec9a3c4a9500 has active endpoints

=== FAIL: github.com/docker/docker/integration/networking TestBridgeICCWindows (16.80s)
```

vvoland commented:
One easy thing to cache would be the `readConfig` calls - no need to repeatedly fetch the same blobs from the content store. Configs are small enough, and with content addressability there's no cache invalidation needed (content won't change for the same digest).
